docs/code-quality-tasks/singleton-patterns-report.md

```diff
@@ -0,0 +1,151 @@
+# Singleton Patterns Report
+
+Based on the analysis of the codebase against AGENTS.md guidelines, the
+following singleton patterns were found that should be refactored:
+
+## 1. Module-level Maps and Caches
+
+### `packages/deno-vite-plugin/src/resolver.ts`
+
+- **Line 75**: `const resolverCache = new Map();`
+- **Issue**: Global cache that prevents multiple instances and complicates
+  testing
+
+### `packages/static/index.ts`
+
+- **Line 5**: `const cache = new StaticCache();`
+- **Issue**: Exported singleton cache instance
+
+## 2. Exported Singleton Instances
+
+### `packages/toolshed/lib/otel.ts`
+
+- **Line 20**: `export const otlpExporter = new OTLPTraceExporter(...)`
+- **Line 26**: `export const provider = new BasicTracerProvider(...)`
+- **Issue**: Global telemetry instances that can't be mocked or isolated for
+  testing
+
+### `packages/toolshed/routes/storage/blobby/utils.ts`
+
+- **Line 7**: `export const storage = new DiskStorage(DATA_DIR);`
+- **Line 10**: `let redisClient: RedisClientType | null = null;`
+- **Issue**: Global storage instance and module-level Redis client state
+
+### `packages/memory/clock.ts`
+
+- **Line 20**: `export default new Clock();`
+- **Issue**: Default export of singleton instance prevents multiple clock
+  instances
+
+## 3. Module-level TextEncoder/TextDecoder Instances
+
+### `packages/identity/src/identity.ts`
+
+- **Line 11**: `const textEncoder = new TextEncoder();`
+
+### `packages/utils/src/encoding.ts`
+
+- **Line 1**: `const encoder = new TextEncoder();`
+- **Line 2**: `const decoder = new TextDecoder();`
+- **Issue**: While less problematic, these could be passed as dependencies for
+  better testability
+
+## 4. Module-level Configuration Objects
+
+### `packages/background-charm-service/src/env.ts`
+
+- **Line 42**: `export const env = loadEnv();`
+- **Issue**: Global environment configuration that can't be easily mocked
```

Comment:
Thoughts on env vars:
* Non-standard! Our modules go through typescript, vite, rollup, deno, npm compat deno, browser. Minimize usage in general.
* Deno services may freely use env vars. Prefer directly using `Deno.env`. Prefer only using in entry file. Quickly represent as a data structure.
* `VITE_` env vars can only be used in jumble. Transitive deps (runner) require non-standard APIs to leverage